### PR TITLE
WIP: Dependency upgrades

### PR DESCRIPTION
**Pull request by fleon**
Dependency upgrades:
- monaco-editor 0.16.2 -> 0.20.0
- monaco-editor-core 0.16.1 -> 0.20.0
- monaco-languages 1.6.0 -> 1.10.0
- monaco-plugin-helpers 1.0.2 -> 1.0.3

Remove dependencies that are also dependencies of yaml-language-server

Remove all code from
- src/languageservice
- src/yaml-ast-parser-custom-tags
- test

Fixes https://github.com/pengx17/monaco-yaml/issues/25

### CODE REVIEW

File: examples/react-webpack-worker-loader/package.json [hunk 1]
```diff
@@ -18,7 +18,7 @@
     "file-loader": "^5.1.0",
     "html-webpack-plugin": "^3.2.0",
     "monaco-editor": "^0.20.0",
-    "monaco-yaml": "^2.4.0",
+    "monaco-yaml": "file:../..",
```
Comment (yazaabed) [inline examples/react-webpack-worker-loader/package.json:21]: Is this means load monaco-yaml from dist? or the main code?

File: package.json [hunk 2]
```diff
@@ -38,33 +38,28 @@
...
   "dependencies": {
-    "js-yaml": "^3.12.0",
-    "vscode-json-languageservice": "^3.4.11",
-    "vscode-languageserver": "^5.2.1",
-    "vscode-uri": "^2.1.1"
+    "yaml-language-server": "^0.9.0"
   },
   "peerDependencies": {
-    "monaco-editor": "^0.19.2"
+    "monaco-editor": "^0.20.0"
```
Comment (yazaabed) [inline package.json:44]: :nice:

File: scripts/bundle-esm.js [hunk 3]
```diff
@@ -3,31 +3,50 @@
...
-const REPO_ROOT = path.join(__dirname, "../");
+const REPO_ROOT = path.join(__dirname, '../');
 
 helpers.packageESM({
   repoRoot: REPO_ROOT,
-  esmSource: "out/esm",
-  esmDestination: "lib/esm",
-  entryPoints: ["monaco.contribution.js", "yamlMode.js", "yaml.worker.js"],
+  esmSource: 'out/esm',
+  esmDestination: 'lib/esm',
```
Comment (yazaabed) [inline scripts/bundle-esm.js:14]: Would it be possible to keep it without esm? Or we are doing the same pattern in other monaco plugins?

File: src/yamlWorker.ts [hunk 4]
```diff
@@ -10,7 +10,7 @@ import Thenable = monaco.Thenable;
 import IWorkerContext = monaco.worker.IWorkerContext;
 
 import * as ls from 'vscode-languageserver-types';
-import * as yamlService from './languageservice/yamlLanguageService';
+import * as yamlService from 'yaml-language-server/out/server/src/languageservice/yamlLanguageService';
```
Comment (yazaabed) [inline src/yamlWorker.ts:13]: :+1:

**Comment by yazaabed** [comment]
I couldn't find the minor version increase?

**Comment by fleon** [comment]
This code doesn't yet work, which is why I put this in WIP. There are issues with Webpack. 

@yazaabed Do you mind trying out the demos and seeing if you can figure out the problem. Webpack says "module not found for a lot of third files in `yaml-language-server`.

File: examples/react-webpack-worker-loader/package.json (hunk 1 above)
Comment (fleon) [reply examples/react-webpack-worker-loader/package.json:21]: From the main code. The 2.4.1 update's structure now allows us to do this.

File: scripts/bundle-esm.js (hunk 3 above)
Comment (fleon) [reply scripts/bundle-esm.js:14]: yes its the same process in other monaco plugins too, except the `lib` folder is called `release`. It is very similar to the [`monaco-json`](https://github.com/microsoft/monaco-json) repo, which this repo is very similar in nature to.

**Comment by fleon** [comment]
> I couldn't find the minor version increase?

What do you mean? The MR doesn't bump up this repo to 2.5.0. It will happen directly on master by running the `yarn publish` command. It automatically creates a version bump commit, a tag and publishes the version to npm.

**Comment by yazaabed** [comment]
> This code doesn't yet work, which is why I put this in WIP. There are issues with Webpack.
> 
> @yazaabed Do you mind trying out the demos and seeing if you can figure out the problem? Webpack says "module not found for a lot of third files in `yaml-language-server`. I think if we just provide the mappings in the script, it should work, but I was not able to get it working that way.

Yeah sure I will clone it today after work and check out why webpack is not working.

**Comment by jjyyxx** [comment]
I need several fixes in the up-to-date `yaml-language-server`, so I looked into the problem.

I made a few changes to get correctly bundled _deps. https://github.com/jjyyxx/monaco-yaml/tree/himkp-housekeeping
 
**NOTE: the solution needs to make a change in `node_modules/monaco-plugin-helpers`**

The root cause is that `monaco-plugin-helpers` is not a full-featured bundler. It relies on `typescript.preProcessFile` to resolve dependencies of a file. `typescript.preProcessFile` by default only support typescript or es6-style imports.

**If you modify `node_modules\monaco-plugin-helpers\lib\packageESM.js#L82` to `var info = ts.preProcessFile(fileContents, true, true);`, the `require` statements will be located.**

Then it almost works, except `preProcessFile` detects dependency `prettier` in `node_modules\yaml-language-server\out\server\src\languageservice\services\yamlFormatter.js`. Use `preProcessFile` on `prettier` is completely broken, so I added a filler to exclude prettier from the optional dependencies. (may have better solutions)

It's not an elegant solution, but at least produces correctly bundled esm/min libraries.

**Comment by fleon** [comment]
@jjyyxx Thanks for the fixes and sorry for the delay in responding here. I was also confused earlier as to why `prettier` was an `optionalDependency` and not a `devDependency`. But it turns out we do need `prettier` in case people want to auto-format the yaml in monaco editor.

Unfortunately, we can't remove `prettier` optional dependency (it would make the library unusable for people relying on the feature of auto-formatting the yaml), and perhaps this is why patching `yaml-language-server` was necessary in the first place. See: https://github.com/pengx17/monaco-yaml/commit/1b31f4818688253ebc897c58f78cfe92cf39adda

I am okay with modifying `monaco-plugin-helpers` in a fork, or to contribute to that upstream repo, but we'll have to figure something about the removal of `prettier`.

Or perhaps I'm not understanding correctly it seems? 🤔 If `prettier` is only removed for the sake of bundling, and can actually be optionally included for people who want to auto-format their YAML, then I am fine with this approach. @jjyyxx Could you please clarify?

**Comment by jjyyxx** [comment]
The function `typescript.preProcessFile`, as its name points out, may use a fast instead of AST-based approach to scan a file. For normal js files, it worked well. But when used with `prettier/standalone.js` or `prettier/index.js`, it produces weird dependencies like `)` and `exports` (with `ts.preProcessFile(fileContents, true, true)` modification).

One workaround would be patching `monaco-plugin-helpers` to 
- add a global option of whether enabling scanning `require` in js files (default to `false` to avoid breaking changes)
- add optional per file based override option, similar to `resolveAlias`, to control a single file's `require` scanning behavior (to workaround tricky dependencies like prettier).

Then it's possible to add a filler like this
```typescript
// throw new Error('Module not included');

import prettier from "prettier/standalone"
import parser from "prettier/parser-yaml"

export default {
    format(text: string, options) {
        prettier.format(
            text,
            Object.assign(options, { plugins: [parser] })
        )
    }
}
```
to avoid patching yaml-language-server to achieve formatting. (currently this would break as the reason mentioned above)

This workaround leaves an edge case.  Files where `require` scanning don't work and actually have `require` dependencies can't be handled by this workaround.

**Comment by yazaabed** [comment]
> The function `typescript.preProcessFile`, as its name points out, may use a fast instead of AST-based approach to scan a file. For normal js files, it worked well. But when used with `prettier/standalone.js` or `prettier/index.js`, it produces weird dependencies like `)` and `exports` (with `ts.preProcessFile(fileContents, true, true)` modification).
> 
> One workaround would be patching `monaco-plugin-helpers` to
> 
> * add a global option of whether enabling scanning `require` in js files (default to `false` to avoid breaking changes)
> * add optional per file based override option, similar to `resolveAlias`, to control a single file's `require` scanning behavior (to workaround tricky dependencies like prettier).
> 
> Then it's possible to add a filler like this
> 
> ```ts
> // throw new Error('Module not included');
> 
> import prettier from "prettier/standalone"
> import parser from "prettier/parser-yaml"
> 
> export default {
>     format(text: string, options) {
>         prettier.format(
>             text,
>             Object.assign(options, { plugins: [parser] })
>         )
>     }
> }
> ```
> 
> to avoid patching yaml-language-server to achieve formatting. (currently this would break as the reason mentioned above)
> 
> This workaround leaves an edge case. Files where `require` scanning don't work and actually have `require` dependencies can't be handled by this workaround.

Could we use dynamic imports from webpack instead of enabling require? Like:

```
export default {
    format(text: string, options) {
         Promise.all([import("prettier/standalone"), import("prettier/parser-yaml")]).then(modules => {
              modules[0].format(
                   text,
                  Object.assign(options, { plugins: [modules[1]] })
               )
         }).catch(error => /* do nothing */);
    }
}

**Comment by jjyyxx** [comment]
My bad, I didn't realize there's an option called `resolveSkip`

Adding 'prettier/standalone', 'prettier/parser-yaml' to `resolveSkip`, use the following prettier filler, and the ESM build finally worked (as a charm).

```typescript
import prettier from 'prettier/standalone'
import parser from 'prettier/parser-yaml'

export function format(text: string, options) {
    return prettier.format(text, Object.assign(options, { plugins: [parser] }))
}
```

Use ES6 dynamic import is another great idea (fixed several mistakes in the original post)
```typescript
export function format(text: string, options) {
    return Promise.all([import('prettier/standalone'), import('prettier/parser-yaml')])
        .then(([prettier, parser]) => prettier.format(text, Object.assign(options, { plugins: [parser] })))
        .catch(error => console.warn(error));
}
```
but **one really tricky thing** is, the YAML LanguageService#doFormat method do not expect a `Promise<string>` result. This promise is then wrapped into a `TextEdit` object, which led to postMessage failure. To handle this, you have to change the `format` method in yamlWorker like this,
```typescript
public format(
  uri: string,
  range: ls.Range,
  options: yamlService.CustomFormatterOptions
): Thenable<ls.TextEdit[]> {
  const document = this._getTextDocument(uri);
  const textEdits = this._languageService.doFormat(document, options);
  return Promise.all(textEdits.map(te => te.newText as never as Promise<string>)).then(values => {
    // actually of 0 or 1 length
    textEdits.forEach((te, i) => te.newText = values[i])
    return textEdits
  })
}
```

I have tested out both snippets above, and can confirm that formatting worked as expected. Feel free to choose one, though I prefer the first method. 

I'm using this package for customized yaml edit service, and made a bunch of modifications and extensions to the source, so I'm not quite ready to clean my code up to make a PR for this task. You could take these pieces of code, and finish this PR.

**Comment by yazaabed** [comment]
> My bad, I didn't realize there's an option called `resolveSkip`
> 
> Adding 'prettier/standalone', 'prettier/parser-yaml' to `resolveSkip`, use the following prettier filler, and the ESM build finally worked (as a charm).
> 
> ```ts
> import prettier from 'prettier/standalone'
> import parser from 'prettier/parser-yaml'
> 
> export function format(text: string, options) {
>     return prettier.format(text, Object.assign(options, { plugins: [parser] }))
> }
> ```
> 
> Use ES6 dynamic import is another great idea (fixed several mistakes in the original post)
> 
> ```ts
> export function format(text: string, options) {
>     return Promise.all([import('prettier/standalone'), import('prettier/parser-yaml')])
>         .then(([prettier, parser]) => prettier.format(text, Object.assign(options, { plugins: [parser] })))
>         .catch(error => console.warn(error));
> }
> ```
> 
> but **one really tricky thing** is, the YAML LanguageService#doFormat method do not expect a `Promise<string>` result. This promise is then wrapped into a `TextEdit` object, which led to postMessage failure. To handle this, you have to change the `format` method in yamlWorker like this,
> 
> ```ts
> public format(
>   uri: string,
>   range: ls.Range,
>   options: yamlService.CustomFormatterOptions
> ): Thenable<ls.TextEdit[]> {
>   const document = this._getTextDocument(uri);
>   const textEdits = this._languageService.doFormat(document, options);
>   return Promise.all(textEdits.map(te => te.newText as never as Promise<string>)).then(values => {
>     // actually of 0 or 1 length
>     textEdits.forEach((te, i) => te.newText = values[i])
>     return textEdits
>   })
> }
> ```
> 
> I have tested out both snippets above, and can confirm that formatting worked as expected. Feel free to choose one, though I prefer the first method.
> 
> I'm using this package for customized yaml edit service, and made a bunch of modifications and extensions to the source, so I'm not quite ready to clean my code up to make a PR for this task. You could take these pieces of code, and finish this PR.

Thanks a lot of trying this out and fix the issue. Really appreciated. @fleon would like to do the last step as @jjyyxx suggested? Or should I do it?

**Comment by yazaabed** [comment]
@fleon can we change this PR from draft to open?

**Comment by dmy147** [comment]
Has there been any progress ?

**Comment by fleon** [comment]
Thanks for working on this @jjyyxx @dmy147 and appreciate the MR. As long as the examples work fine, we should be good with merging this MR now. 

@yazaabed Could you please take a look at it once again? Thanks!

**Comment by jjyyxx** [comment]
just to be clear, I didn't test or fix the dev/umd bundle

**Review by yaabed — APPROVED** [review]

**Review by yazaabed — APPROVED** [review]

**Comment by fleon** [comment]
Thanks for the contributions everyone. I will create a new version 2.5.0 for this soon.

**Comment by yazaabed** [comment]
@fleon sadly I didn't have prettier enabled when I tested the final result on my project. Which is throwing an error because of this line https://github.com/redhat-developer/yaml-language-server/blob/c1e6f58da43c0f041c0e0afd3c7739193fc71269/src/languageservice/services/yamlFormatter.ts#L27 inside the yaml-language-server.

**Comment by yazaabed** [comment]
@fleon sadly I didn't have prettier enabled when I tested the final result on my project. Which is throwing an error because of this line https://github.com/redhat-developer/yaml-language-server/blob/c1e6f58da43c0f041c0e0afd3c7739193fc71269/src/languageservice/services/yamlFormatter.ts#L27 inside the yaml-language-server.

Error:
```
ModuleNotFoundError: Module not found: Error: Can't resolve 'fs' in '/webapp/node_modules/monaco-yaml/node_modules/vscode-nls/lib'
```

**Comment by jjyyxx** [comment]
> Which is throwing an error because of this line https://github.com/redhat-developer/yaml-language-server/blob/c1e6f58da43c0f041c0e0afd3c7739193fc71269/src/languageservice/services/yamlFormatter.ts#L27 inside the yaml-language-server.

This error is weird, The error seems related to `vscode-nls` which has been substituted with

```
'vscode-nls': path.join(REPO_ROOT, 'out/esm/fillers/vscode-nls.js'),
```

**Comment by yazaabed** [comment]
> > Which is throwing an error because of this line https://github.com/redhat-developer/yaml-language-server/blob/c1e6f58da43c0f041c0e0afd3c7739193fc71269/src/languageservice/services/yamlFormatter.ts#L27 inside the yaml-language-server.
> 
> This error is weird, The error seems related to `vscode-nls` which has been substituted with
> 
> ```
> 'vscode-nls': path.join(REPO_ROOT, 'out/esm/fillers/vscode-nls.js'),
> ```

Yeah I though the same, but when I changed manually the yaml-language-server/yamlFormatter by hand and wrote prettier/singleton it worked fine.

**Comment by jjyyxx** [comment]
> I need several fixes in the up-to-date `yaml-language-server`, so I looked into the problem.
> 
> I made a few changes to get correctly bundled _deps. https://github.com/jjyyxx/monaco-yaml/tree/himkp-housekeeping
> 
> **NOTE: the solution needs to make a change in `node_modules/monaco-plugin-helpers`**
> 
> The root cause is that `monaco-plugin-helpers` is not a full-featured bundler. It relies on `typescript.preProcessFile` to resolve dependencies of a file. `typescript.preProcessFile` by default only support typescript or es6-style imports.
> 
> **If you modify `node_modules\monaco-plugin-helpers\lib\packageESM.js#L82` to `var info = ts.preProcessFile(fileContents, true, true);`, the `require` statements will be located.**
> 
> Then it almost works, except `preProcessFile` detects dependency `prettier` in `node_modules\yaml-language-server\out\server\src\languageservice\services\yamlFormatter.js`. Use `preProcessFile` on `prettier` is completely broken, so I added a filler to exclude prettier from the optional dependencies. (may have better solutions)
> 
> It's not an elegant solution, but at least produces correctly bundled esm/min libraries.

I mentioned before that you have to make a change in `monaco-plugin-helpers`, did you miss that?

**Comment by yazaabed** [comment]
> > I need several fixes in the up-to-date `yaml-language-server`, so I looked into the problem.
> > I made a few changes to get correctly bundled _deps. https://github.com/jjyyxx/monaco-yaml/tree/himkp-housekeeping
> > **NOTE: the solution needs to make a change in `node_modules/monaco-plugin-helpers`**
> > The root cause is that `monaco-plugin-helpers` is not a full-featured bundler. It relies on `typescript.preProcessFile` to resolve dependencies of a file. `typescript.preProcessFile` by default only support typescript or es6-style imports.
> > **If you modify `node_modules\monaco-plugin-helpers\lib\packageESM.js#L82` to `var info = ts.preProcessFile(fileContents, true, true);`, the `require` statements will be located.**
> > Then it almost works, except `preProcessFile` detects dependency `prettier` in `node_modules\yaml-language-server\out\server\src\languageservice\services\yamlFormatter.js`. Use `preProcessFile` on `prettier` is completely broken, so I added a filler to exclude prettier from the optional dependencies. (may have better solutions)
> > It's not an elegant solution, but at least produces correctly bundled esm/min libraries.
> 
> I mentioned before that you have to make a change in `monaco-plugin-helpers`, did you miss that?

Yeah, but after that we discussed that `resolveSkip` fixed that issue.

**Comment by jjyyxx** [comment]
Not really

You have to change `var info = ts.preProcessFile(fileContents);` to `var info = ts.preProcessFile(fileContents, true, true);`. Otherwise, `require` will not be recognized as an import (and left as is) , thus the filler won't work.

**Comment by yazaabed** [comment]
> Not really
> 
> You have to change `var info = ts.preProcessFile(fileContents);` to `var info = ts.preProcessFile(fileContents, true, true);`. Otherwise, `require` will not be recognized as an import (and left as is) , thus the filler `fillers/prettier.ts` won't work.

Looks like @dmy147 did these changes. Check the screenshot please. 
Screen Shot 2020-08-20 at 1 10 05 PM

**Comment by yazaabed** [comment]
> > Not really
> > You have to change `var info = ts.preProcessFile(fileContents);` to `var info = ts.preProcessFile(fileContents, true, true);`. Otherwise, `require` will not be recognized as an import (and left as is) , thus the filler `fillers/prettier.ts` won't work.
> 
> Looks like @dmy147 did these changes. Check the screenshot please.
> Screen Shot 2020-08-20 at 1 10 05 PM

@jjyyxx weird the change is inside a patch file. Is this correct? Or something wrong with the other PR?

**Comment by jjyyxx** [comment]
`patch-package` was good, I used it in another project.

And I found the problem. Instead of
```
resolveSkip: ['monaco-editor', 'monaco-editor-core', 'js-yaml', 'prettier'],
```
you should write
```
resolveSkip: [
    'monaco-editor',
    'monaco-editor-core',
    'js-yaml',
    'prettier/standalone',
    'prettier/parser-yaml',
],
```

Skip `prettier` led to the error.

**Comment by fleon** [comment]
Yes it is fine to have that change in a patch file, and in fact I liked the solution a lot. Since we don't have ownership of the target repository, and we don't really want to create and maintain our own fork of that repo, patching it is the easiest and least hacky way to solve the problem.

@jjyyxx Thanks for figuring out the issue quickly. I have opened https://github.com/pengx17/monaco-yaml/pull/30 to resolve it.

**Pull request merged**